proposals/infra/INF-0005-Long-Vector-TestPlan.md

```diff
@@ -0,0 +1,273 @@
+mm_annotate_shader_op_arith_table.py (WinTools repo) will need to be updated to
+recognize any new tests (not test cases) added. And additional GUIDs added for
+new test cases. mm_annotate_shader_op_arith_table.py is called by
+mm-hlk-update.py when converting from ExecutionTests to the HLK 'DxilConf' tests.
+The aforementioned *table.py script is run by Integration.HLKTestsUpdate.yaml
+
+# Test Validation Requirements
+The following statements must be true and validated for this work to be
+considered completed.
+- All new test cases pass when run locally against a WARP device
+- All new test cases are confirmed to be lit up and available in the HLK when
+  the target device reports support
+- All new tests/test cases are added to the HLK playlist
+
+# Requirements
+- Greg's long vector changes:
+  https://github.com/microsoft/hlsl-specs/blob/main/proposals/0026-hlsl-long-vector-type.md#allowed-elementwise-vector-intrinsics
+- WARP long vector support (Jesse). Needs Greg's work, or a private WIP
+   branch of Greg's work. ETA of ~1 week to implement.
+
+# Notes
+- Private test binaries/collateral will be shared with IHVs for validation
+   purposes. This will enable IHVs to verify long vector functionality without
+   waiting for an OS/HLK release.
+
+# Open Questions
+- We speculate that the combination of test cases for HLSL operators and HLSL
+intrinsics which map to DXIL ops should get us all, or most of the coverage
+required for the 'emulated' intrinsics. But we should audit that.
+
+# HLSL Operators
+These operators should generate LLVM native ops which use vectors.
+
+Operator table from [Microsoft HLSL Operators](https://learn.microsoft.com/en-us/windows/win32/direct3dhlsl/dx-graphics-hlsl-operators)
+| Operator Name | Operator | Notes |
+|-----------|--------------|----------|
+Additive and Multiplicative Operators | +, -, *, /, % |
+Array Operator | [i] | llvm:ExtractElementInst
+Assignment Operators | =, +=, -=, *=, /=, %= |
+Bitwise Operators | ~, <<, >>, &, \|, ^, <<=, >>=, &=, \|=, ^= | Only valid on
+||| int and uint vectors
+Boolean Math Operators | & &, ||, ?: |
+Cast Operator | (type) | No direct operator, difference in GetElementPointer 
+||| or load type
+Comparison Operators| <, >, ==, !=, <=, >= |
+Prefix or Postfix Operators| ++, -- |
+Unary Operators | !, -, + |
+
+# Mappings of HLSL intrinsics to DXIL OPcodes or LLVM Operators
+
+Note: [ ] - Brackets are used in the below tables to signify that the operator
+is only used in specific paths. Operators sharing the same brackers are in the
+same logic path. If an operator is not in brackets then it is used in all cases.
+
+## Trigonometry
+
+| Intrinsic | DXIL OPCode | LLVM Operator | Notes |
+|-----------|--------------|----------|-----------|
+| acos      | Acos | | |
+| asin      | Asin | | |
+| atan      | Atan | | |
+| atan2     | Atan | CreateFDiv, CreateFAdd, CreateFSub, CreateFCmpOLT,
+||| CreateFCpmOEQ, CreateFCmpOGE, CreateFCmpOLT, CreateAnd, CreateSelect | |
```

Comment:
I'm guessing that these "CreateXXX" function names came from the C++ code.  The LLVM instructions are things like `fdiv`, `fadd`, `fcmp`, `and`, `select` etc.

I also note that if I try `atan2` in godbolt it doesn't actually generate `fsub`. https://godbolt.org/z/7xqz6448P

Reply:
Yes, they're the C++ helper functions. I was intending to distill down further on a subsequent iteration. Wanted to sanity check that they do what I would expect (always return the llvm instruction in the name).

I'll have to peel a little more in TranslateAtan2, for this 'inventory' I went through the various Translate* functions in HLSLOperatorLower and noted the possible operations and instructions. For [TranslateAtan2 it looks like FSub](https://github.com/microsoft/DirectXShaderCompiler/blob/9e8a698deed37116e5e55cebd0d725c5c2be5e4c/lib/HLSL/HLOperationLower.cpp#L1849) is always used, so it must be optimized or simplified out along the way?

Reply:
Asked for some help. The FSub is optimized out to normalize to FAdd. If you pass '-O0' instead you'll see that the FSub is there.

But this makes me wonder about the right testing approach to ensure coverage. Given that the goal is to test that DXIL Ops and LLVM Instructions are functional with vectors, maybe we want to disable optimizations for the test cases?